docs/testing.md

```diff
@@ -0,0 +1,156 @@
+
+It is possible that a test is needed to learn how the interaction between the operator and the dependency should play out in the context of the given specification.
+
+But **do not assume the test built will become the final e2e test** for the change or feature PR.
+
+We will refer to this testing phase as _exploratory testing_.
+
+### Prefer TDD
+
+**When the detailed behavior on part of the operator is clear, you can start writing the unit or integration test that covers it**. A unit test is preferable, as it will allow faster and more reliable development feedback loops. 
+
+**Keep your tests focused on a single check**. Note a single _exploratory test_ scenario will usually imply several interactions to be tested individually.
+
+**Try to write the test before**, following the specification and any details learnt from the exploratory testing, if any. Test Driven Development (TDD) will force you to think on the spec and the interface of the change or feature design before its implementation. It will probably inform you to use more arguments and simpler functions so you can test the specified behavior more easily in isolation.
+
+**Avoid testing implementation details**, focus on specified outcomes and behaviors. [More on this later](#prefer-external-tests)
+
+Testing against a real Kubernetes, Atlas or a Cloud Provider is not possible on a unit test. But we have support for mocking those dependencies sufficiently so that we can test most of the operator behavior without resorting to an e2e test or integration test. Those support mechanisms are explained in the [Test isolation support section](#test-isolation-support).
+
+### Evaluate additional testing requirements
+
+Once unit tests work as expected, you should be pretty confident your code does what the test spec says. Still you might want to add an integration test if the feature covers a complex workflow or set of steps.
+
+It might even be the case now that the initial exploratory testing code must be converted to an e2e test. Note this should not be done lightly, as e2e tests add costs to the development process and CI pipeline feedback. If unit tests and/or integration tests cover the change or feature properly and the e2e feels redundant, it is preferably to not include it. On the other hand, if the e2e might catch issues that could have been missed otherwise, or need to detect dependency behavior changes, you should include the e2e, for those specific reasons, but not as the primary means to cover the change or feature testing.
+
+### <a name="prefer-external-tests"></a>Prefer external tests
+
+When writing tests for a feature or change, it is best if those can exercise the code as an external consumer, without any knowledge of or access to our code internal implementation. This makes for tests on behavior rather than implementation details. In Go this usually means prefer `package xxx_test` rather than `package xxx` directly.
```

Comment:
This may make unit tests more complicated and harder to write. I agree we should test the behaviour, but I wouldn't ban using the same package name

Reply:
This is not a ban. It is a reminder of a best practice. Our code right now breaks this rule more often than it should, most probably due to writing the test after the implementation. But it is good to remember, when possible, it is best to test from the outside using the intended external interfaces.

See https://jdkaplan.dev/thinkin-logs/2021-10-07/ although not proposing about internal packages, at least not yet.

In other words, when there is no other way, just use `package xxx`, when creating new features, try TDD (write the test first) and use the external interface, so you should be able to test from `package xxx_test`.